sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala

```diff
@@ -0,0 +1,127 @@
+
+      /*
+          // Add @a as an element before each element of @as
+          private def intersperse[A](a: A, as: List[A]): List[A] = as match {
+            case Nil    => Nil
+            case h :: t => List(a, h) ++ intersperse(a, t)
+          }
+
+          val headers = reqStub.headers.toList.flatMap {
+              case (k, v) if k != "Origin" => List(k, v)
+              // We need to unpack multiple Origins in discrete key-value pairs
+              // to play nicely with [HttpRequest.builder().headers()]
+              case (k, v) => intersperse(k, v.split(",").toList)
+            }
+       */
+
+      val address = Uri.fromString(properties.endpoint + generatedRequest.method.path.toString()) match {
+        case Right(value) => value
+        case Left(_) =>
+          throw new Exception(
+            "Error: cannot produce uri from endpoint provided: %s , with path %s"
+              .format(properties.endpoint, generatedRequest.method.path.toString)
+          )
+      }
+      val body = generatedRequest.body match {
+        case Some(b) => b.toString()
+        case _       => ""
+      }
+
+      val req = generatedRequest.method match {
+
+        // POST requests: attach body, use the uri we already have, without adding querysring
+        case trackerMethod.Post(_) =>
+          Request[F](method = Method.POST, uri = address).withEntity(body).putHeaders(headers)
+
+        // GET requests: add querystring, ignore body field
+        case trackerMethod.Get(_) =>
+          val q = generatedRequest.qs.getOrElse("").toString()
+          val newAddress = Uri.fromString(address.toString() + "?" + q) match {
+            case Right(value) => value
+            case Left(_) =>
+              throw new Exception(
+                "Error: Could not append querystring: %s , to endpoint %s"
+                  .format(generatedRequest.qs.toString(), address.toString)
+              )
+          }
+          Request[F](method = Method.GET, uri = newAddress).putHeaders(headers)
+
+        // HEAD requests: gave headaches so just sending simple hardcoded get requests instead for now.
+        case _ =>
+          val dummyAddress = Uri.fromString("http://localhost:9090/i?e=pv").getOrElse(throw new Exception("borked"))
+          // MAking empty get requests as HEADs were throwing errors.
+          // TODO: Figure this out
+          Request[F](method = Method.GET, uri = dummyAddress)
+      }
+      // TODO: Can this be cleaner by just pattern matching the assignment of method, querystring and body?
+      // If a get request can have WithEntity(""), then it might work
+
+      return req
+    }
+
+    val httpClient = EmberClientBuilder.default[F].build
+
+    st: Stream[F, Main.GenOutput] =>
+      st.map(_._3).map(buildRequesst).evalMap(req => httpClient.use(client => client.expect[String](req))).void
```

Comment:
Here when you call `httpClient.use { ..... }`  it creates and destroys the http client + all of the internal objects needed to implement it.  Like a TCP connection pool etc.

The problem is, you call `.use` for every single request.  So you're continually creating/destroying the resources, which is quite wasteful.

It's better to find a way to `.use` the http client just once, and then re-use it for every request.  I'd be happy to show you a few ways to arrange the code for this.